### DISCUSSION TITLE
Number formatting

### DISCUSSION THREAD

**Discussion by olazzzzz** [freeplane/freeplane#1928] [Help] [answered]
Hi everyone,

Pretty basic question, but I can't seem to make this work.
I have to work with big numbers on my mindmaps, and I would like to have them shown in this format: 1.000.000.000,87 
So points to separate the thousands and a comma to separate the decimals. 
Can I achieve this in the number format settings?
Thanks in advance for your help.

**Comment by carrots5G** [comment]
Sorry I cannot help, but actually I would also be interested in knowing how to format numbers in this way

**Reply by euu2021** [reply]
Try those options:

Preferences…->Environment->Language->Use Operating System Regional Settings

Preferences…->Behaviour->Data formatting and parsing->Standard number format

Preferences…->Behaviour->Data formatting and parsing->Locale for formats

Also: https://github.com/freeplane/freeplane/discussions/769

**Comment by olazzzzz** [comment]
@euu2021 Thanks for your input. I am not sure exactly how I should change those options to fit my needs.
I think the related topic also is different from what I am looking for.
Basically, if you have node A = 1.000.000.000, and node B = 4,365, I would like that node C is able to calculate their quotient and express it in this format, C =229.095,074; whereby the dots separate the thousands and commas the decimals. 
Any help is appreciated!

**Reply by euu2021** [reply]
So, do you want to have a formula/script that takes the content of node A and the content of node B and calculate their quotient? I think this can be done using the basic ideal behind the scripts at #769, isn't it?

I think I need another clarification, please. In the node core, do you write the numbers like this:

1000000000

Or like this:

1.000.000.000

**Comment by olazzzzz** [comment]
@euu2021 Thank you for your reply and willingness to help!

If a script is the way to go then great, until now I was just trying to do it via the "edit formula" (see image), seemed to me visually better to have this way. In this example I'm trying to divide the first node by the second.

![image](https://github.com/user-attachments/assets/8f947bde-9815-4438-bb6e-a14986f9e9fe)

It would be preferable for my purposes to already use the dots and commas in the node core of the first node, but if I can input it without dots and commas and the result would have them, that's also satisfactory for me.

**Reply by euu2021** [reply, marked as answer]
I see. You can write the numbers in your preferred format, and you can use a formula. Let's see:

What happens is that, in order to make the math operation, you need to make Java understand that the text that you have in your node is a number object. 

So, in my example, those will be nodes A and B:

```
nodeA = ID_1504482427
nodeB = ID_510131125
```

Now, we take the text in nodes A and B, and parse it into a number:

```
numberA = NumberFormat.getNumberInstance(Locale.GERMANY).parse(nodeA.text)
numberB = NumberFormat.getNumberInstance(Locale.GERMANY).parse(nodeB.text)
```

Then we make the math operation:

```
quotient = numberA/numberB
```

Finally, we format the final result to the desired style:

```
NumberFormat.getNumberInstance(Locale.GERMANY).format(quotient)
```

This is the complete formula:

```
import java.text.NumberFormat
import java.util.Locale

nodeA = ID_1504482427
nodeB = ID_510131125

numberA = NumberFormat.getNumberInstance(Locale.GERMANY).parse(nodeA.text)
numberB = NumberFormat.getNumberInstance(Locale.GERMANY).parse(nodeB.text)

quotient = numberA/numberB

NumberFormat.getNumberInstance(Locale.GERMANY).format(quotient)
```

![image](https://github.com/user-attachments/assets/6bafaa00-b91f-412f-ad2e-ddcf2cf39af8)

**Comment by olazzzzz** [comment]
Thank you @euu2021 , that is precisely what I needed!
One last question, how would you add a line to the formula so that the result is rounded to the second decimal place?
I have been trying for one hour, the best I could reach was to cut on the second decimal, i.e., it's not rounding the last digit.
Thanks for your patience and help, very appreciated!

**Reply by euu2021** [reply]
To get serious about rounding, let's use BigDecimal, because it allows choosing many kinds of roundings.

The script gets one more step, like this:

```
import java.text.NumberFormat
import java.util.Locale
import java.math.BigDecimal
import java.math.RoundingMode

nodeA = ID_1504482427
nodeB = ID_510131125

numberA = NumberFormat.getNumberInstance(Locale.GERMANY).parse(nodeA.text)
numberB = NumberFormat.getNumberInstance(Locale.GERMANY).parse(nodeB.text)

quotient = numberA/numberB

BigDecimal roundedNumber = new BigDecimal(quotient.toString()).setScale(2, RoundingMode.HALF_UP)

NumberFormat.getNumberInstance(Locale.GERMANY).format(roundedNumber)
```

In `setScale(2, RoundingMode.HALF_UP)` the 2 is the number of decimals, and RoundingMode.HALF_UP is one of the modes that you can choose ([see more here](https://docs.oracle.com/javase/8/docs/api/java/math/RoundingMode.html)). This is a summary of the different rounding modes:


![image](https://github.com/user-attachments/assets/1404df04-7fb5-42ee-ae02-2695883caa8b)

**Comment by olazzzzz** [comment]
Fantastic! Thank you so much @euu2021, you are the best!
Truly appreciated!
